File: 31-next-permutation/next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int pivot = -1;
        int n = nums.size();
        for(int i = n - 1; i > 0; i--) {
            if(nums[i] > nums[i - 1]) {
                pivot = i - 1;
                break;
            }
        }

        if(pivot == -1) {
            reverse(nums.begin(), nums.end());
            return;
        }
        reverse(nums.begin() + pivot + 1, nums.end());
        for(int i = pivot + 1; i < n; i++) {
            if(nums[i] > nums[pivot]) {
                swap(nums[i], nums[pivot]);
                break;
            }
        }
    }
};

File: 31-next-permutation/next-permutation.cs
public class Solution {
    public void NextPermutation(int[] nums) {
        int n = nums.Length;
        int pivot = -1;
        for(int i = n - 2; i >= 0; i--) {
            if(nums[i] < nums[i + 1]) {
                pivot = i;
                break;
            }
        }
        if(pivot == -1) {
            Array.Reverse(nums);
            return;
        }
        Array.Reverse(nums, pivot + 1, n - pivot - 1);
        int idx = -1;
        for(int i = pivot + 1; i < n; i++) {
            if(nums[i] > nums[pivot]) {
                (nums[i], nums[pivot]) = (nums[pivot], nums[i]);
                return;
            }
        }

    }
}